python/ray/autoscaler/_private/fake_multi_node/test_utils.py

```diff
@@ -240,3 +245,66 @@ def stop(self):
             f"RAY_FAKE_CLUSTER=1 ray down -y {self.config_file}", shell=True)
 
         self._stop_monitor()
+
+    def _update_nodes(self):
+        with open(self._nodes_file, "rt") as f:
+            self._nodes = json.load(f)
+
+    def _update_status(self):
+        with open(self._status_file, "rt") as f:
+            self._status = json.load(f)
+
+    def _get_node(self,
+                  node_id: Optional[str] = None,
+                  num: Optional[int] = None,
+                  rand: Optional[str] = None) -> str:
+        self._update_nodes()
+        if node_id:
+            assert not num and not rand, (
+                "Only provide either `node_id`, `num`, or `random`.")
+        elif num:
+            assert not node_id and not rand, (
+                "Only provide either `node_id`, `num`, or `random`.")
+            base = "fffffffffffffffffffffffffffffffffffffffffffffffffff"
+            node_id = base + str(num).zfill(5)
+        elif rand:
+            assert not node_id and not num, (
+                "Only provide either `node_id`, `num`, or `random`.")
+            assert rand in ["worker", "any"
+                            ], ("`random` must be one of ['worker', 'any']")
+            choices = list(self._nodes.keys())
+            if rand == "worker":
+                choices.remove(
+                    "fffffffffffffffffffffffffffffffffffffffffffffffffff00000")
+            # Else: any
+            node_id = random.choice(choices)
+
+        assert node_id in self._nodes, (
+            f"Node with ID {node_id} is not in active nodes.")
+        return node_id
+
+    def _get_docker_container(self, node_id: str) -> Optional[str]:
+        self._update_status()
+        node_status = self._status.get(node_id)
+        if not node_status:
+            return None
+
+        return node_status["Name"]
+
+    def kill_node(self,
```

Comment:
Is this supposed to be called from main test thread?

See  a try I did (https://github.com/ray-project/ray/pull/21516/commits/d139750dcbe1f93f758d4a2e458777993fdb167a). The main test thread is occupied by tune.run when we want to kill a docker node. I used another process to kill it. I wonder if I could leverage this API somehow.

Reply:
The way we currently do it in other tests is using callbacks, would that work for you?